### ISSUE TITLE
trivy-db fails to download with  UNAUTHORIZED: The client does not have permission for manifest aquasecurity/trivy-db/2/manifest.json

### ISSUE THREAD

**Issue by seb-835** [aquasecurity/trivy-operator#695]
**What steps did you take and what happened:**

trivy-operator pod failed to update db with error : 

```
init error: DB error: failed to download vulnerability DB:
 OCI artifact error: OCI artifact error: OCI repository error: GET https://******/v2/aquasecurity/trivy-db/manifests/2: UNAUTHORIZED: The client does not have permission for manifest; map[manifest:aquasecurity/trivy-db/2/manifest.json]\n"
,"stacktrace":"github.com/aquasecurity/trivy-operator/pkg/vulnerabilityreport.(*WorkloadController).processFailedScanJob\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/vulnerabilityreport/controller.go:551\ngithub.com/aquasecurity/trivy-operator/pkg/vulnerabilityreport.(*WorkloadController).reconcileJobs.func1\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/vulnerabilityreport/controller.go:376\nsigs.k8s.io/controller-runtime/pkg/reconcile.Func.Reconcile\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/reconcile/reconcile.go:102\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:121\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:320\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234"}  
```

**What did you expect to happen:**

I expected the trivy-db to be downloaded fine.

**Anything else you would like to add:**

What permission do i have to look for and authorized to fix this issue ?  
Thanks a lot for advice. 

**Environment:**

- Trivy-Operator version : 0.33.0
- Helm-chart install :   trivy-operator-0.6.0
- Kubernetes version : 1;24.4
- OS  : Redhat 7.9

**Comment by chen-keinan** [comment]
@seb-835  you should not set up a permission for it , is this happen intermittently? or happen on every scan job?

**Comment by seb-835** [comment]
@chen-keinan it happens for every scan job. None Start at all.

**Comment by chen-keinan** [comment]
@seb-835 we have not changed anything about it in `trivy-operator v0.6.0` can you check if you get the same issue with `trivy-operator v0.5.0`

**Comment by chen-keinan** [comment]
> @chen-keinan i install trivy-operator v0.5.0, and got same behaviour.

thanks for checking , I think it might be env. / config  issue.
can you please share more info on env./ special config you are using ?

**Comment by chen-keinan** [comment]
if you do not have an access to `(ghcr.io)` , then this is your problem as the `trivy-db` now is only hosted on `(ghcr.io)`

**Comment by seb-835** [comment]
ok, i will fix that on the registry to get the image from ghcr.io
and get back to you if i got the same behaviour

**Comment by seb-835** [comment]
@chen-keinan : is it expected ? 

```
docker pull ghcr.io/aquasecurity/trivy-db:2
2: Pulling from /aquasecurity/trivy-db:2
b4f49697b38B : Pulling fs Layer
unexpected end of Json Input
```
```
crictl  --config=/var/lib/rancher/rke2/agent/etc/crictl.yaml pull ghcr.io/aquasecurity/trivy-db:2

E1116 14:45:30.116981   75414 remote_image.go:238] "PullImage from image service failed" err="rpc error: code = InvalidArgument desc = failed to pull and unpack image \"ghcr.io/aquasecurity/trivy-db:2\": unable to fetch descriptor (sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855) which reports content size of zero: invalid argument" image="ghcr.io/aquasecurity/trivy-db:2"

FATA[0001] pulling image: rpc error: code = InvalidArgument desc = failed to pull and unpack image "ghcr.io/aquasecurity/trivy-db:2": unable to fetch descriptor (sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855) 
which reports content size of zero: invalid argument
```

**Comment by chen-keinan** [comment]
> @chen-keinan : is it expected ?
> 
> ```
> docker pull ghcr.io/aquasecurity/trivy-db:2
> 2: Pulling from /aquasecurity/trivy-db:2
> b4f49697b38B : Pulling fs Layer
> unexpected end of Json Input
> ```

you need to use `oras` see [docs](https://github.com/aquasecurity/trivy-db)

**Comment by seb-835** [comment]
Thanks for help @chen-keinan 
now, all images are pull from ghcr.io, but still having the UNAUTHORIZED issue.

**Comment by chen-keinan** [comment]
> Thanks for help @chen-keinan 
> now, all images are pull from ghcr.io, but still having the UNAUTHORIZED issue.

Is it the sma error or different, can you share the log ?

**Comment by seb-835** [comment]
Same error  

`{"level":"error","ts":1668610817.513167,"logger":"reconciler.vulnerabilityreport","msg":"Scan job container","job":"trivy-system/scan-vulnerabilityreport-7759bd949","container":"79ffd7da-935c-43f3-9c1b-0c0c6fdf9fc3","status.reason":"Error","status.message":"2022-11-16T15:00:14.258Z\t\u001b[34mINFO\u001b[0m\tNeed to update DB\n2022-11-16T15:00:14.258Z\t\u001b[34mINFO\u001b[0m\tDB Repository: registry.xxxx/aquasecurity/trivy-db\n2022-11-16T15:00:14.258Z\t\u001b[34mINFO\u001b[0m\tDownloading DB...\n2022-11-16T15:00:14.631Z\t\u001b[31mFATAL\u001b[0m\tinit error: DB error: failed to download vulnerability DB: OCI artifact error: OCI artifact error: OCI repository error: GET https://registry.xxxx/v2/aquasecurity/trivy-db/manifests/2: UNAUTHORIZED: The client does not have permission for manifest; map[manifest:aquasecurity/trivy-db/2/manifest.json]\n","stacktrace":"github.com/aquasecurity/trivy-operator/pkg/vulnerabilityreport.(*WorkloadController).processFailedScanJob\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/vulnerabilityreport/controller.go:551\ngithub.com/aquasecurity/trivy-operator/pkg/vulnerabilityreport.(*WorkloadController).reconcileJobs.func1\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/vulnerabilityreport/controller.go:376\nsigs.k8s.io/controller-runtime/pkg/reconcile.Func.Reconcile\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile/reconcile.go:102\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:121\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:320\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.0/pkg/internal/controller/controller.go:234"}`

Does the job use the same ServiceAccount as trivy-operator deployment  to get the imagePullSecret to get the trivy-db ?

**Comment by chen-keinan** [comment]
Nope, It doesn't use secret , it assume  its public

**Comment by seb-835** [comment]
hum,   even if my artifactory registry pull the image from ghcr.io as puvlic  , my cluster need to authenticate against the artifactory registry  to get the image.  

anyway, if you do not specify a service account in the job manifest,  it may use the "default" one of the trivy-system namespace,
it put the imagePullSecret  in the default too, but still the same issue :(

**Comment by chen-keinan** [comment]
Download db is done in init container need to check if we can pass TRIVY_USER/TRIVY_PASSWORD    then it will work

**Comment by chen-keinan** [comment]
There is a helm param:
`Trivy.imagePullSecret` could you set it up with private registry secret

**Comment by chen-keinan** [comment]
@seb-835 I will try to find a away to make it work with current release  if I will not succeed I  will create PR for it

**Comment by chen-keinan** [comment]
@seb-835  yes it do create [entry](https://github.com/aquasecurity/trivy-operator/blob/main/deploy/helm/templates/config.yaml#L76) in trivy-operator-trivy-config cm.

still not sure it will solve you problem , I need to review it deeply and get back to you.

**Comment by seb-835** [comment]
Sorry for late, yes, they are.

**Comment by chen-keinan** [comment]
> Sorry for late, yes, they are.

we might need to do a change in `trivy` to support download db from private registry , I will get back to you with final answer shortly

**Comment by mfilotto** [comment]
Any update on this one please ?

**Comment by chen-keinan** [comment]
> Any update on this one please ?

I'll ooen an issue on trivy to support download db (Server) from  private registry

**Comment by sathieu** [comment]
> > Any update on this one please ?
> 
> I'll open an issue on trivy to support download db (Server) from private registry

This is already possible:

https://github.com/aquasecurity/trivy-operator/blob/487faa08eaac282e4b4018e8dea4b426a6fe0445/deploy/helm/values.yaml#L339

**Comment by chen-keinan** [comment]
> This is already possible:
> 
> https://github.com/aquasecurity/trivy-operator/blob/487faa08eaac282e4b4018e8dea4b426a6fe0445/deploy/helm/values.yaml#L339

this is for public repo not private , if it require credential it will now work

**Comment by sathieu** [comment]
Thanks for the clarification.

**Comment by chen-keinan** [comment]
> @chen-keinan what is the number of the issue you opened on trivy ?

thanks for reminder , here is the [issue](https://github.com/aquasecurity/trivy/issues/3891)

**Comment by mfilotto** [comment]
@chen-keinan the [PR](https://github.com/aquasecurity/trivy/pull/3915) for this issue on trivy has just been merged, and the issue  is closed.

**Comment by chen-keinan** [comment]
> @chen-keinan the [PR](https://github.com/aquasecurity/trivy/pull/3915) for this issue on trivy has just been merged, and the issue is closed.

@mfilotto great. once trivy v0.39.0 will be released I'll update `trivy-operator`

**Comment by mfilotto** [comment]
@chen-keinan: how can I configure authentication in helm chart ?

**Comment by chen-keinan** [comment]
> @chen-keinan: how can I configure authentication in helm chart ?

[User](https://github.com/aquasecurity/trivy-operator/blob/main/deploy/helm/values.yaml#L371) and [Password](https://github.com/aquasecurity/trivy-operator/blob/main/deploy/helm/values.yaml#L373)

**Comment by mfilotto** [comment]
@chen-keinan I created a secret named trivy-operator-trivy-config with values TRIVY_USERNAME and TRIVY_PASSWORD, but I still have the error UNAUTHORIZED in trivy operator logs.
This secret seems to be used only in builtInTrivyServer mode, should I actiavte that mode ?
I'm not sure to understand what this mode is for though

**Comment by chen-keinan** [comment]
> @chen-keinan I created a secret named trivy-operator-trivy-config with values TRIVY_USERNAME and TRIVY_PASSWORD, but I still have the error UNAUTHORIZED in trivy operator logs. This secret seems to be used only in builtInTrivyServer mode, should I actiavte that mode ? I'm not sure to understand what this mode is for though

yes , what I have specified is for `builtInTrivyServer`.
`builtInTrivyServer` param is simply spin up in cluster server, it remove the server concern from the user

**Comment by mfilotto** [comment]
Now I have this error in trivy server logs
 TRIVY_USERNAME and TRIVY_PASSWORD are correctly set in secret trivy-operator-trivy-config
---
$ kubectl logs -f trivy-server-0
2023-04-05T23:17:27.933Z        WARN    'TRIVY_SKIP_UPDATE' is deprecated. Use 'TRIVY_SKIP_DB_UPDATE' instead.
2023-04-05T23:17:27.934Z        INFO    Need to update DB
2023-04-05T23:17:27.934Z        INFO    DB Repository: docker-nexus.sln.nc/aquasecurity/trivy-db
2023-04-05T23:17:27.934Z        INFO    Downloading DB...
2023-04-05T23:17:27.988Z        FATAL   failed to download vulnerability DB: database download error: OCI repository error: 1 error occurred:
        * GET https://docker-nexus.sln.nc/v2/aquasecurity/trivy-db/manifests/2: UNAUTHORIZED: access to the requested resource is not authorized

**Comment by chen-keinan** [comment]
> Now I have this error in trivy server logs TRIVY_USERNAME and TRIVY_PASSWORD are correctly set in secret trivy-operator-trivy-config
> 
> $ kubectl logs -f trivy-server-0 2023-04-05T23:17:27.933Z WARN 'TRIVY_SKIP_UPDATE' is deprecated. Use 'TRIVY_SKIP_DB_UPDATE' instead. 2023-04-05T23:17:27.934Z INFO Need to update DB 2023-04-05T23:17:27.934Z INFO DB Repository: docker-nexus.sln.nc/aquasecurity/trivy-db 2023-04-05T23:17:27.934Z INFO Downloading DB... 2023-04-05T23:17:27.988Z FATAL failed to download vulnerability DB: database download error: OCI repository error: 1 error occurred: * GET https://my-private-registry/v2/aquasecurity/trivy-db/manifests/2: UNAUTHORIZED: access to the requested resource is not authorized

@mfilotto Thanks for feedback,  I see now that the [PR](https://github.com/aquasecurity/trivy/pull/3915) done on trivy support only `trivy image ...` sub command  and not `trivy server...` example:
```shell
$ TRIVY_USERNAME=foo TRIVY_PASSWORD=bar trivy image --db-repository ghcr.io/your/private-db --download-db-only
```

I'm Checking it with trivy project and getting back to you , opening this issue for now

**Comment by chen-keinan** [comment]
@mfilotto @seb-835 the latest PR in trivy is not supported for server command.
I have created another [PR-Server support](https://github.com/aquasecurity/trivy/pull/4008) to support server side , once merged and tagged , I'll be able to update trivy version

**Comment by chen-keinan** [comment]
@mfilotto now adding trivy v0.40.0 which include credential server support for private repo, will be available with trivy-operator v0.13.1
